### PR TITLE
[arrayAUC] Fix tie handling issue

### PR DESCRIPTION
**Pull request by gabrielmcg44**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix tie handling in `arrayAUC` to match sklearn

Previously, when 2 labels had the same scores, it considered that each label would result in a new ROC point.  This was leading to differences between Clickhouse AUC and market standard sklearn.

Now, a new point is only considered when the score changes. What might result in a trapezoidal area being added. This results in AUCs identical to sklearn.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings:
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit 80c489041ef7e6d52fc00d9f5ec1922f1d7f505d with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65840/80c489041ef7e6d52fc00d9f5ec1922f1d7f505d/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/2673a773c36f5e12b45e8c3b9b97b6cc3703afc1/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by gabrielmcg44** [comment]
To illustrate the issue, i created a simple example with tied scores:
scores = [0.1, 0.35, 0.4, 0.8, 0.8]
labels = [0, 1, 0, 1, 0]

First, I tried it using python sklearn (result = 0.58333333334):

auc-sklearn

After it, I tried with clickhouse with the order of the labels just like I showed here (result = 0.666666667):

auc-clickhouse-1

And then I changed the order of the labels with score 0.8 (result = 0.5):

auc-clickhouse-2

After  this change, I ran clickhouse locally and tried both ordered and they matched sklearn result:

Screenshot from 2024-06-28 09-41-20

I also tried with a real case where I first noticed the divergences in my job, and the results matched sklearn's

**Comment by den-crane** [comment]
* https://github.com/ClickHouse/ClickHouse/discussions/65192

**Comment by gabrielmcg44** [comment]
I opened to review, but I did not manage to make all tests pass yet. For example some stateless tests failed in CI, but are passing when I run them locally. I would really appreciate some hints

**Comment by alexey-milovidov** [comment]
> Bugfix validation — New test(s) failed to reproduce a bug 

It looks like the new test also passed before this PR.

**Comment by gabrielmcg44** [comment]
> > Bugfix validation — New test(s) failed to reproduce a bug
> 
> It looks like the new test also passed before this PR.

This is awkward.. I tried this by downloading the executable and ran the test pointing to a server created by it, and it is failing like this
Screenshot from 2024-06-29 15-14-51

**Comment by gabrielmcg44** [comment]
> > > Bugfix validation — New test(s) failed to reproduce a bug
> > 
> > 
> > It looks like the new test also passed before this PR.
> 
> This is awkward.. I tried this by downloading the executable and ran the test pointing to a server created by it, and it is failing like this Screenshot from 2024-06-29 15-14-51

Also tried with master version and got the same error

**Comment by CurtizJ** [comment]
> > Bugfix validation — New test(s) failed to reproduce a bug

> It looks like the new test also passed before this PR.

Bugfix validation check was broken #65946.

**Review by CurtizJ — APPROVED** [review]

**Comment by CurtizJ** [comment]
Test failures are unrelated.

**Comment by gabrielmcg44** [comment]
Thanks @CurtizJ ! Just to understand, this change will be available on the next release, right?

**Comment by CurtizJ** [comment]
Yes, in 24.7 release.
